googlecartographer/cartographer
# Do not install test libraries (libgmock* / libgtest*)

Issue
====
Cartographer installs custom gtest and gmock libraries and headers.


I face this issue recently when packaging cartographer in ROS. cartographer currently installs gtest and gmock library in the global library directory. This makes other packages fail as they will find these custom libraries rather than the system ones.
This problem happens only on "recent" operating systems (Debian Stretch and Ubuntu Artful and Bionic).
More details available on https://github.com/ros-gbp/cartographer-release/issues/8

I'm opening this PR for feedback mostly about the approach and will remove the commented code if the maintainers decide this is the right approach.

Nature of the change:
=================
- modify the macro `_common_compile_stuff` to respect the passed `VISIBILITY` argument
- modify the macro `google_test` to make the linkage of the gmock libraries private
- modify the main CMakeLists.txt to not add the GMock includes and libraries as part of the main target.

How to test the patch:
 ===========
Requirement:
- a machine running one of the following OSs: Debian Stretch (9), Ubuntu Artful (17.10) or Ubuntu Bionic (18.04)
- the packages `python-rosdep` and `ros-lunar-catkin` installed from the ROS repositories

To save time I use ROS docker images:
`docker run -it --rm osrf/ros:lunar-desktop-full-stretch`

1. Get the code with the patch
```
mkdir -p /tmp/catkin_ws/src && cd /tmp/catkin_ws/src
git clone https://github.com/mikaelarguedas/cartographer -b dont_install_test_libs
```
2. Edit package.xml to use the ceres rosdep key
```
cd /tmp/catkin_ws/src/cartographer
sed -i 's/depend>ceres-solver</depend>libceres-dev</g' package.xml
cd /tmp/catkin_ws
```
3. Get the dependencies
```
rosdep update
rosdep install --from-paths src --ignore-src --rosdistro lunar -y
```
4. Build cartographer
```
source /opt/ros/lunar/setup.bash
catkin_make_isolated --install
```
here libgtest* and libgmock* should NOT be listed in the installation step:
```
ls install_isolated/lib/
libcartographer.a
```

5. Run the tests:
```
source install_isolated/setup.bash
catkin_make_isolated --install --make-args test
```
All tests should run and pass.
